presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayNormalizeFunction.java

```diff
@@ -0,0 +1,138 @@
+    @TypeParameterSpecialization(name = "T", nativeContainerType = double.class)
+    @SqlType("array(T)")
+    @SqlNullable
+    public static Block normalizeDoubleArray(
+            @TypeParameter("T") Type elementType,
+            @SqlType("array(T)") Block block,
+            @SqlType("T") double p)
+    {
+        return normalizeArray(elementType, block, p, DOUBLE_VALUE_ACCESSOR);
+    }
+
+    @TypeParameter("T")
+    @TypeParameterSpecialization(name = "T", nativeContainerType = long.class)
+    @SqlType("array(T)")
+    @SqlNullable
+    public static Block normalizeRealArray(
+            @TypeParameter("T") Type elementType,
+            @SqlType("array(T)") Block block,
+            @SqlType("T") long p)
+    {
+        return normalizeArray(elementType, block, Float.intBitsToFloat((int) p), REAL_VALUE_ACCESSOR);
+    }
+
+    private static Block normalizeArray(Type elementType, Block block, double p, ValueAccessor valueAccessor)
+    {
+        if (!(elementType instanceof RealType) && !(elementType instanceof DoubleType)) {
+            throw new PrestoException(
+                    FUNCTION_IMPLEMENTATION_MISSING,
+                    format("Unsupported array element type for array_normalize function: %s", elementType.getDisplayName()));
+        }
+        checkCondition(p >= 0, INVALID_FUNCTION_ARGUMENT, "array_normalize only supports non-negative p: %s", p);
+
+        if (p == 0) {
+            return block;
+        }
+
+        int elementCount = block.getPositionCount();
+        double pNorm = 0;
+        for (int i = 0; i < elementCount; i++) {
+            if (block.isNull(i)) {
+                return null;
+            }
+            pNorm += Math.pow(Math.abs(valueAccessor.getValue(elementType, block, i)), p);
+        }
+        if (pNorm == 0) {
+            return block;
+        }
+        pNorm = Math.pow(pNorm, 1.0 / p);
+        BlockBuilder blockBuilder = elementType.createBlockBuilder(null, elementCount);
+        for (int i = 0; i < elementCount; i++) {
+            valueAccessor.writeValue(elementType, blockBuilder, valueAccessor.getValue(elementType, block, i) / pNorm);
+        }
+        return blockBuilder.build();
+    }
+
+    private interface ValueAccessor
+    {
+        double getValue(Type elementType, Block block, int position);
+
+        void writeValue(Type elementType, BlockBuilder blockBuilder, double value);
+    }
+
+    private static class DoubleValueAccessor
+            implements ValueAccessor
+    {
+        @Override
+        public double getValue(Type elementType, Block block, int position)
+        {
+            return elementType.getDouble(block, position);
+        }
+
+        @Override
+        public void writeValue(Type elementType, BlockBuilder blockBuilder, double value)
+        {
+            elementType.writeDouble(blockBuilder, value);
+        }
+    }
+
+    private static class RealValueAccessor
+            implements ValueAccessor
+    {
+
```

Comment:
this empty line is failing the style checking I think

Reply:
Removed. Interesting. `Format file` in IntelliJ doesn't fix it.